### PR TITLE
[doc] Add docs for push-based shuffle in Datasets

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Adds recommendations, example, and brief benchmark results for push-based shuffle in Datasets.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/performance-tips.rst [hunk 1]
```diff
@@ -122,3 +122,49 @@ If you're seeing out of memory errors during map tasks, reducing the max block s
...
+These operations include ``Dataset.random_shuffle``, ``Dataset.sort`` and ``Dataset.groupby``.
+Shuffle can be challenging to scale to large data sizes and clusters, especially when the total dataset size cannot fit into memory.
+
+Starting in Ray v1.13, Datasets provides an alternative shuffle implementation known as push-based shuffle for improving large-scale performance.
+We recommend trying this out if your dataset has more than 1k partitions (input files) or 1TB of data.
+
+To try this out locally or on a cluster, you can start with the `nightly release test <https://github.com/ray-project/ray/blob/master/release/nightly_tests/dataset/sort.py>`_ that Ray runs for ``Dataset.random_shuffle`` and ``Dataset.sort``.
+To get an idea of the performance you can expect, here are some run time results for ``Dataset.random_shuffle`` on 1-10TB of data on 20 machines (m5.4xlarge instances on AWS EC2, each with 16 vCPUs, 64GB RAM).
+
+.. image:: https://docs.google.com/spreadsheets/d/e/2PACX-1vQvBWpdxHsW0-loasJsBpdarAixb7rjoo-lTgikghfCeKPQtjQDDo2fY51Yc1B6k_S4bnYEoChmFrH2/pubchart?oid=598567373&format=image
```
Comment (scv119) [inline doc/source/data/performance-tips.rst:139]: should we compare the push based vs the default one in the graph to demonstrate it's more appealing.?

**Review by scv119 — APPROVED** [review]
Looks great!

File: doc/source/data/performance-tips.rst (hunk 1 above)
Comment (stephanie-wang) [reply doc/source/data/performance-tips.rst:139]: I tried to run the default one but it failed to complete heh... I could maybe run it on a smaller data size? But we won't be able to provide reliable comparison points past 1TB.

Comment (scv119) [reply doc/source/data/performance-tips.rst:139]: ah it's tricky then

**Pull request merged**